packages/fxa-auth-server/lib/routes/recovery-phone.ts

```diff
@@ -522,6 +522,144 @@ class RecoveryPhoneHandler {
     throw AppError.invalidOrExpiredOtpCode();
   }
 
+  async replacePhoneNumber(request: AuthRequest) {
+    // need to check first that there is an existing phone number
+    const { uid, email } = request.auth
+      .credentials as SessionTokenAuthCredential;
+
+    const { code } = request.payload as unknown as {
+      code: string;
+    };
+
+    //guard against invalid token
+    if (!email) {
+      throw AppError.invalidToken();
+    }
+
+    // check if user has a confirmed recovery phone number
+    const { exists } = await this.recoveryPhoneService.hasConfirmed(uid);
+    const codeIsValid = await this.recoveryPhoneService.validateSetupCode(
+      uid,
+      code
+    );
+
+    if (!exists) {
+      throw AppError.recoveryPhoneNumberDoesNotExist();
+    }
+
+    if (!codeIsValid) {
+      // if the code is invalid, we need to throw so that we don't remove their old
+      // phone before we confirm the new one.
+      throw AppError.invalidOrExpiredOtpCode();
+    }
+
+    let removedSuccess = false;
+    try {
+      removedSuccess = await this.recoveryPhoneService.removePhoneNumber(uid);
+      if (removedSuccess) {
+        recordSecurityEvent('account.recovery_phone_removed', {
+          db: this.db,
+          request,
+        });
+      }
+    } catch (error) {
+      if (error instanceof RecoveryNumberRemoveMissingBackupCodes) {
+        throw AppError.recoveryPhoneRemoveMissingRecoveryCodes();
+      }
+
+      throw AppError.backendServiceFailure(
+        'RecoveryPhoneService',
+        'removePhoneNumber',
+        { uid },
+        error
+      );
+    }
+
+    if (!removedSuccess) {
+      // removing the first phone failed and we can't proceed
+      throw AppError.backendServiceFailure(
+        'RecoveryPhoneService',
+        'replacePhoneNumber',
+        { uid },
+        new Error('Failed to remove existing phone number')
+      );
+    }
+
+    // setup new phone number recovery with code provided
+    let confirmNewSuccess = false;
+    try {
+      confirmNewSuccess = await this.recoveryPhoneService.confirmSetupCode(
+        uid,
+        code
+      );
+    } catch (error) {
+      if (error instanceof RecoveryPhoneNotEnabled) {
+        throw AppError.featureNotEnabled();
+      } else if (error instanceof RecoveryNumberNotExistsError) {
+        throw AppError.recoveryPhoneNumberDoesNotExist();
+      } else if (error instanceof RecoveryNumberAlreadyExistsError) {
+        throw AppError.recoveryPhoneNumberAlreadyExists();
+      } else {
+        throw AppError.backendServiceFailure(
+          'RecoveryPhoneService',
+          'confirmCode',
+          { uid },
+          error
+        );
+      }
+    }
+
+    if (!confirmNewSuccess) {
+      // if we get here, the code was invalid, but we removed the existing phone number.
```

Comment:
Is there a reason we can't confirm the code first, then remove, then replace? I think that order of operations would avoid the issue you bring up here, which seems like a valid concern to me.